src/vision/status_quo/alan_thinks_he_needs_async_locks.md

```diff
@@ -0,0 +1,121 @@
+  }
+}
+```
+
+Alan is happy that his code seems to be compiling! The short but hard learning curve has been worth it. He's having fun now!
+
+Unfortunately, Alan's happiness soon comes to end as he starts integrating his request handler into calls to `tokio::spawn` which he knows will allow him to manage multiple requests at a time. The error message is somewhat cryptic, but Alan is confident he'll be able to figure it out:
+
+```
+189 |     tokio::spawn(async {
+    |     ^^^^^^^^^^^^ future created by async block is not `Send`
+::: /home/alan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.5.0/src/task/spawn.rs:129:21
+    |
+129 |         T: Future + Send + 'static,
+    |                     ---- required by this bound in `tokio::spawn`
+
+note: future is not `Send` as this value is used across an await
+   --> src/handler.rs:787:9
+      |
+786   |         let routes = self.state.lock().unwrap();
+      |             - has type `std::sync::MutexGuard<'_, HashMap<String, Url>>` which is not `Send`
+787   |         Ok(self.client.perform_request(route, request).await?)
+      |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `routes` maybe used later
+788   |     })
+      |     - `routes` is later dropped here
+```
+
+Alan stops and takes a deep breath. He tries his best to make sense of the error message. He sort of understands the issue the compiler is telling him. Apparently `routes` is not marked as `Send`, and because it is still alive over a call to `await`, it is making the future his handler returns not `Send`. And tokio's `spawn` function seems to require that the future it received be `Send`. 
+
+Alan reaches the boundaries of his knowledge of Rust, so he reaches out over chat to ask his co-worker Barbara for help. Not wanting to bother her, Alan provides the context he's already figured out for himself.
+
+Barbara knows that mutex guards are not `Send` because sending mutex guards to different threads is not a good idea. She suggests looking into async locks which can be held across await points because they are `Send`. Alan looks into the tokio documentation for more info and is easily able to move the use of the standard library's mutex to tokio's mutex. It compiles!
+
+Alan ships his code and it gets a lot of usage. After a while, Alan notices some potential performance issues. It seems his proxy handler does not have the throughput he would expect. Barbara, having newly joined his team, sits down with him to take a look at potential issues. Barbara is immediately worried by the fact that the lock is being held much longer than it needs to be. The lock only needs to be held while accessing the route and not during the entire duration of the downstream request.
+
+She suggests to Alan to switch to not holding the lock across the I/O operations. Alan first tries to do this by explicitly cloning the url and dropping the lock before the proxy request is made:
```

Comment:
I think we are missing a transition _back_ from async to non-async locks somewhere.

Reply:
They do that later. Perhaps that's not clear?

Reply:
But they are already back in the snippet just below my comment. Otherwise they wouldn't get the error.